### PR TITLE
ber_metaocaml: (re)init at 104

### PR DESCRIPTION
**Pull request by mpickering**
###### Motivation for this change

The old derivation was broken as the `src` pointed to a mutable url. It was also necessary to make some more packaging changes to get it to compile to follow with changes to the normal ocaml build process.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by mpickering** [comment]
Possible reviewers? @vbgl @aske @Zimmi48  ?

I am not an ocaml expert but this appears to work for me.

File: pkgs/development/compilers/ocaml/ber-metaocaml-104.nix [hunk 1]
```diff
@@ -7,52 +7,70 @@ let
...
-    url = "http://okmij.org/ftp/ML/ber-metaocaml.tar.gz";
-    sha256 = "1kq1if25c1wvcdiy4g46xk05dkc1am2gc4qvmg4x19wvvaz09gzf";
+    url = "http://okmij.org/ftp/ML/ber-metaocaml-104.tar.gz";
+    sha256 = "1gmwlxairxqcmqa2r6kbf8b4dxc7pfhfbh48g1s14d3z20rj8nib";
   };
 
   # Needed to avoid a SIGBUS on the final executable on mips
   NIX_CFLAGS_COMPILE = if stdenv.isMips then "-fPIC" else "";
 
   patches = optionals stdenv.isDarwin [ ./gnused-on-osx-fix.patch ];
```
Comment (vbgl) [inline pkgs/development/compilers/ocaml/ber-metaocaml-104.nix]: This patch does not apply. Is it needed?

File: pkgs/development/compilers/ocaml/ber-metaocaml-104.nix [hunk 2]
```diff
@@ -7,52 +7,70 @@ let
...
-  buildInputs = [ncurses] ++ optionals useX11 [ xlibsWrapper ];
+  configureFlags = optionals useX11 [ "-x11lib" x11lib
+                                      "-x11include" x11inc ];
+
+  dontStrip = true;
+  buildInputs = [ncurses] ++ optionals useX11 [ libX11 xproto ];
   installFlags = "-i";
   installTargets = "install"; # + optionalString useNativeCompilers " installopt";
   prePatch = ''
     CAT=$(type -tp cat)
```
Comment (vbgl) [inline pkgs/development/compilers/ocaml/ber-metaocaml-104.nix]: Is this `CAT` variable used anywhere? If not, you may drop the full `prePatch` attribute.

**Review by vbgl — CHANGES_REQUESTED** [review]

**Comment by mpickering** [comment]
@vbgl as a generic point if I have some more time in the future would it be better to modify `generic.nix` to also allow building `ber_metaocaml`? There is quite a lot of duplicated logic as it is essentially an ocaml build with some additional patches.

**Comment by mpickering** [comment]
Thanks for the sharp eye @vbgl , I removed the two redundant parts of the expression.

**Review by vbgl — APPROVED** [review]

**Comment by vbgl** [comment]
Thanks. Your suggestion to refactor `generic.nix` makes sense.

**Pull request merged**